src/main/c/NativeModule.cpp

```diff
@@ -0,0 +1,625 @@
+  key_(key),
+  is_global_(is_global),
+  dlopen_handle_(nullptr),
+  lib_name_(config.get_lib_name(key_)),
+  new_name_(config.get_new_name(key_)) {
+  std::lock_guard<std::mutex> mylock(big_mutex);
+  // Worker constructor - try to get the binary written to local file
+  if (is_global_) return;
+  int rc = 0;
+  config.ensure_module_dir_exists();
+  struct stat lib_stat;
+  if (file_stat(lib_name_, &lib_stat) && (lib_stat.st_size == binary_size)) {
+    build_state_ = kPass;
+  } else {
+    // We hold big_mutex so there is no race
+    int fd = open(new_name_.c_str(), O_WRONLY|O_CREAT|O_TRUNC, 0666);
+    if (fd < 0) {
+      perror("open");
+      assert(0);
+    }
+    // Now we're about to write the new file
+    rc = write(fd, binary, binary_size);
+    assert(rc == binary_size);
+    ::close(fd);
+    ::chmod(new_name_.c_str(), 0644);
+    struct stat st;
+    if (file_stat(lib_name_, &st)) {
+      long old_size = st.st_size;
+      if (old_size == binary_size) {
+        ::unlink(new_name_.c_str());
+      } else if (old_size == 0) {
+        ::unlink(lib_name_.c_str());
+      } else {
+        auto old = lib_name_ + ".old";
+        ::rename(lib_name_.c_str(), old.c_str());
+      }
+      // Don't let anyone see the file until it is completely written
+      rc = ::rename(new_name_.c_str(), lib_name_.c_str());
+      build_state_ = ((rc == 0) ? kPass : kFail);
+    }
+  }
+  if (build_state_ == kPass) {
+    try_load_locked();
+  }
+}
+
+NativeModule::~NativeModule() {
+  if (!is_global_ && dlopen_handle_) {
+    dlclose(dlopen_handle_);
+  }
+}
+
+bool NativeModule::try_wait_for_build_locked() {
+  if (build_state_ == kInit) {
+    // Since build or write-binary is performed synchronously while holding
+    // big_mutex, and we are holding big_mutex, we know that neither is 
+    // happening.  The lib_file may or may not exist: we just have to
+    // update the state of this NativeModule to match the filesystem.
+    build_state_ = (file_exists(lib_name_) ? kPass : kFail);
+    if (build_state_ == kFail) {
+      std::string base(config.module_dir_ + "/hm_" + key_);
+      fprintf(stderr, "makefile:\n%s", read_file_as_string(base+".mak").c_str());
+      fprintf(stderr, "errors:\n%s",   read_file_as_string(base+".err").c_str());
+    }
+  }
+  return (build_state_ == kPass);
+}
+
+bool NativeModule::try_load_locked() {
+  if (load_state_ == kInit) {
+    if (is_global_) {
+      load_state_ = kPass;
+    } else if (!try_wait_for_build_locked()) {
+      load_state_ = kFail;
+    } else {
+      auto handle = dlopen(lib_name_.c_str(), RTLD_GLOBAL|RTLD_NOW);
+      if (!handle) {
+        fprintf(stderr, "ERROR: dlopen failed: %s\n", dlerror());
+      }
+      load_state_ = (handle ? kPass : kFail);
+      if (handle) dlopen_handle_ = handle;
+    }
+  }
+  return (load_state_ == kPass);
+}
+
+std::vector<char> NativeModule::get_binary() {
+  std::lock_guard<std::mutex> mylock(big_mutex);
+  try_wait_for_build_locked();
```

Comment:
If `try_wait_for_build_locked` goes away (because the build state is resolved in the constructor, you won't need to lock here anymore.

Reply:
True *if* we enforce that a lib file is immutable once it's written (which seems to be where
the other issues are pointing).